source/Test.cpp

```diff
@@ -0,0 +1,186 @@
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+*/
+
+#include "Test.h"
+
+#include "DataNode.h"
+#include "Files.h"
+#include "GameData.h"
+#include "MainPanel.h"
+#include "Panel.h"
+#include "Planet.h"
+#include "PlanetPanel.h"
+#include "PlayerInfo.h"
+#include "Ship.h"
+#include "System.h"
+#include "TestData.h"
+#include "UI.h"
+
+#include <SDL2/SDL.h>
+
+#include <algorithm>
+#include <map>
+#include <stdexcept>
+
+using namespace std;
+
+
+
+namespace{
+	const map<Test::Status, const string> STATUS_TO_TEXT = {
+		{Test::Status::ACTIVE, "active"},
+		{Test::Status::KNOWN_FAILURE, "known failure"},
+		{Test::Status::MISSING_FEATURE, "missing feature"}
+	};
+	
+	const map<Test::TestStep::Type, const string> STEPTYPE_TO_TEXT = {
+		{Test::TestStep::Type::ASSIGN, "assign"},
+		{Test::TestStep::Type::ASSERT, "assert"},
+		{Test::TestStep::Type::BRANCH, "branch"},
+		{Test::TestStep::Type::INJECT, "inject"},
+		{Test::TestStep::Type::INPUT, "input"},
+		{Test::TestStep::Type::INVALID, "invalid"},
+		{Test::TestStep::Type::LABEL, "label"},
+		{Test::TestStep::Type::NAVIGATE, "navigate"},
+		{Test::TestStep::Type::WATCHDOG, "watchdog"}
+	};
+}
+
+
+Test::TestStep::TestStep(Type stepType) : stepType(stepType)
+{
+};
+
+
+
+void Test::LoadSequence(const DataNode &node)
+{
+	for(const DataNode &child: node)
+	{
+		const string &typeName = child.Token(0);
+		Test::TestStep::Type stepType;
+		auto it = find_if(STEPTYPE_TO_TEXT.begin(), STEPTYPE_TO_TEXT.end(),
+			[&typeName](const std::pair<Test::TestStep::Type, string> &e) {
+				return e.second == typeName;
+			});
+		if(it != STEPTYPE_TO_TEXT.end())
+			stepType = it->first;
+		else
+			child.PrintTrace("Unknown teststep type " + child.Token(0));
+		steps.emplace_back(stepType);
+	}
+}
+
+
+
+void Test::Load(const DataNode &node)
+{
+	if(node.Size() < 2)
+	{
+		node.PrintTrace("No name specified for test");
+		return;
+	}
+	// If a test object is "loaded" twice, that is most likely an error (e.g.
+	// due to a plugin containing a test with the same name as the base game
+	// or another plugin). Tests should be globally unique.
+	if(!name.empty())
+	{
+		node.PrintTrace("Duplicate definition of test");
+		return;
+	}
+	// Validate if the testname contains valid characters.
+	if(node.Token(1).find_first_not_of("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789 _-") != std::string::npos)
+	{
+		node.PrintTrace("Invalid test name");
+		return;
+	}
+	name = node.Token(1);
+
+	for(const DataNode &child : node)
+	{
+		if(child.Token(0) == "status" && child.Size() >= 2)
+		{
+			const string &statusText = child.Token(1);
+			auto it = find_if(STATUS_TO_TEXT.begin(), STATUS_TO_TEXT.end(),
+				[&statusText](const std::pair<Test::Status, string> &e) {
+					return e.second == statusText;
+				});
+			if(it != STATUS_TO_TEXT.end())
+				status = it->first;
+			else
+				child.PrintTrace("Unknown test-status " + statusText);
```

Comment:
We may want to assign the status to something other than the default in this scenario, given that the default is "active"

(running `--tests` will display the "unknown status" message, then the table of tests will report it as "active")